File: libs/power-log-parser/src/lib/parsers/total-magnetize-changed-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Node, NodeType, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class TotalMagnetizeChangedParser implements ActionParser {
	readonly ParserName = 'TotalMagnetizeChangedParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, facade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = facade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.BACON_NUM_MAGNETIZE_THIS_GAME as number)
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		if (tagChange.Entity !== this.StateFacade.LocalPlayer!.Id) {
			return null;
		}

		const newValue = tagChange.Value;
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'TOTAL_MAGNETIZE_CHANGED',
				GameEventHelper.CreateProvider(
					'TOTAL_MAGNETIZE_CHANGED',
					null as any,
					this.StateFacade.LocalPlayer!.PlayerId,
					this.StateFacade.LocalPlayer!.Id,
					this.StateFacade,
					{
						NewValue: newValue,
					},
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/TotalMagnetizeChangedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class TotalMagnetizeChangedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public TotalMagnetizeChangedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            TagChange tagChange = null;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.BACON_NUM_MAGNETIZE_THIS_GAME;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities.GetValueOrDefault(tagChange.Entity);

            var controller = entity.GetController();
            if (controller != StateFacade.LocalPlayer.PlayerId)
            {
                return null;
            }

            var newValue = tagChange.Value;
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                 "TOTAL_MAGNETIZE_CHANGED",
                GameEvent.CreateProvider(
                    "TOTAL_MAGNETIZE_CHANGED",
                    null,
                    entity.GetEffectiveController(),
                    entity.Id,
                    StateFacade,
                    new {
                        NewValue = newValue,
                    }),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
